### PR TITLE
feat: lay foundations for engine drivers

### PR DESCRIPTION
**Pull request by jedevc**
This patch starts to simplify the connhelpers and docker-image helper into specific engine drivers (see https://github.com/dagger/dagger/issues/5583) - adding new drivers now becomes easier.

Each driver registers itself under a name, and is called when connecting to a URL with that scheme. Additionally, during connection, we provide a set of additional parameters.

This is mostly just a refactor around the connhelpers, which we can't use directly out of the box, since we want to support a set of additional parameters as well. Since each driver just returns a net.Conn, then we can build the buildkit connection on top of it, which allow us to potentially change the connection mechanism in the future.

### CODE REVIEW

**Review by TomChv — APPROVED** [review]
Do we have any tests on our CI that tests these different drivers?

I like the new interface style registration tho, look pretty cool!

**Comment by gerhard** [comment]
Do we want to continue this?

**Comment by jedevc** [comment]
`TestModuleLotsOfFunctions` is timing-out, looks legitimate :cry:

**Comment by jedevc** [comment]
The `test` failure is happening on `main`, so not relevant to these changes likely. They first occurred after merging https://github.com/dagger/dagger/pull/6597.

**Review by TomChv — APPROVED** [review]

**Comment by jedevc** [comment]
Let's take this now - if it breaks things, would rather do it now, instead of waiting for the big zenith release :smile: 

Also, this has become a lot more relevant as we've been discussing #5583

**Pull request merged**
